content/blog/2020-09-25-efficient-path-profiling/index.md

```diff
@@ -0,0 +1,85 @@
+
+### 2. Optimizing weight selection
+
+The weights selected in step 1 could have suboptimal overhead. They are not selected with minimizing the number of weights in mind. Ideally we'd want fewer edges to have weights, so we have to do less computation when we're profiling a program execution. The next step of the algorithm helps to alleviate this by using [some related work on program event counting][eventcounting]. At a high level, the algorithm finds a maximum cost spanning tree, where the cost is the sum of the weights found in the first part of the algorithm. Each chord $c$ of this spanning tree is then assigned a new weight computed from summing the weights from the first part of the algorithm found in a cycle that contains $c$, the source and the sink of the graph. This transformation also preserves the properties we wanted in the first part of the algorithm, that is, that each path starting at $v$ still maps to a unique number in the range $0..\text{NUMPATHS}(v)$.
+
+### 3. Instrumenting code
+
+The algorithm augments the code we're profiling so that when we run the code, we have logic that tracks which edges are traversed and which weights are added to our running sum. 
+
+### Regenerating paths
+
+Finally, the paper provides a straightforward algorithm to recover a path from a number.
+
+1. Initialize $R$ to have the generated path number, and $v$ to be the entry vertex of the CFG. Keep track of the starting vertex in $\text{PATH}$.
+2. Find edge $v \rightarrow w$ that has the largest $\text{VAL}(v \rightarrow w)$ that's less than $R$. Set $R$ to $R-\text{VAL}(v \rightarrow w)$, and set $v$ to $w$. Append $v$ to $\text{PATH}$.
+3. Stop when you end up at the sink. $\text{PATH}$ will contain a list of vertices.
+
+## Augmenting CFGs with cycles
+
+The algorithm described in this paper only works on directed acyclic graphs (DAGs). The paper describes a technique for modifying the CFG to transform it into a DAG but still provide accurate path profiling, which is crucial to profile programs with loops.
+
+## Evaluation
+
+The authors ran their path profiling implementation on a variety of benchmarks, as well as edge profiling to get a sense of the relative overhead and accuracy. In general, path profiling performance tended to be comparable to edge profiling, and even when benchmarks were large and required more memory intensive instrumentation for path profiling, the overhead was tolerably worse than that of edge profiling.
```

Comment:
A little more detail would be useful here too, in the empirical evaluation. You can apply the things we talked about in the first day of 6120, when we read the performance analysis paper, to use a skeptical lens to interpret the evaluation. How satisfying was the set of benchmarks? Did they measure the right thing? Did they adequately control for potential confounding factors? How solid was the statistical analysis? (For old papers, many of these judgments need to be "graded on a curve," but it's worth pondering nonetheless.) What questions are you left with?

Reply:
Added some critical examination of the evaluation.